File: SenLib/Sen1/FileFixes/scripts_scena_dat_us_t0050_dat.cs
using HyoutaUtils;
using System.Collections.Generic;
using System.IO;

namespace SenLib.Sen1.FileFixes {
	public class scripts_scena_dat_us_t0050_dat : FileMod {
		public string GetDescription() {
			return "Fix minor formatting issues in the Gymnasium.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var s = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xd363fc2114ec8421ul, 0xc24b47c29a4a2badul, 0xed31cfb5u));
			if (s == null) {
				return null;
			}
			MemoryStream bin = s.CopyToMemoryAndDispose();

			// double space
			var patcher = new SenScriptPatcher(bin);
			patcher.RemovePartialCommand(0x11b02, 0x44, 0x11b10, 1);

			// formatting issues in Laura Chapter 1 Day bonding event
			bin.SwapBytes(0x14e92, 0x14e95);
			bin.SwapBytes(0x1505a, 0x15061);
			bin.SwapBytes(0x15186, 0x1518e);
			bin.SwapBytes(0x153a1, 0x153a6);

			// formatting issues in Laura Chapter 4 Evening bonding event
			bin.SwapBytes(0x162a6, 0x162ab);
			bin.SwapBytes(0x164ae, 0x164b4);
			bin.SwapBytes(0x164d8, 0x164dd);
			bin.SwapBytes(0x16627, 0x1662a);

			// formatting issues in Laura Chapter 5 Day bonding event
			bin.Position = 0x17070;
			bin.WriteUInt8(0x01);

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_us/t0050.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var s = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xd363fc2114ec8421ul, 0xc24b47c29a4a2badul, 0xed31cfb5u));
			if (s == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_us/t0050.dat", s) };
		}
	}
}

File: native/sen1/file_fixes/scripts_scena_dat_us_t0050_dat.cpp
#include <string_view>
#include <vector>

#include "sen/file_getter.h"
#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen/sen_script_patcher.h"
#include "sha1.h"
#include "util/memwrite.h"

namespace SenLib::Sen1::FileFixes::scripts_scena_dat_us_t0050_dat {
std::string_view GetDescription() {
    return "Fix formatting issues in the Gymnasium.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_us/t0050.dat",
            221932,
            SenPatcher::SHA1FromHexString("d363fc2114ec8421c24b47c29a4a2baded31cfb5"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        using namespace HyoutaUtils::MemWrite;

        // double space
        SenScriptPatcher patcher(bin);
        patcher.RemovePartialCommand(0x11b02, 0x44, 0x11b10, 1);

        // formatting issues in Laura Chapter 1 Day bonding event
        std::swap(bin[0x14e92], bin[0x14e95]);
        std::swap(bin[0x1505a], bin[0x15061]);
        std::swap(bin[0x15186], bin[0x1518e]);
        std::swap(bin[0x153a1], bin[0x153a6]);

        // formatting issues in Laura Chapter 4 Evening bonding event
        std::swap(bin[0x162a6], bin[0x162ab]);
        std::swap(bin[0x164ae], bin[0x164b4]);
        std::swap(bin[0x164d8], bin[0x164dd]);
        std::swap(bin[0x16627], bin[0x1662a]);

        // formatting issues in Laura Chapter 5 Day bonding event
        WriteUInt8(&bin[0x17070], 0x01);

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen1::FileFixes::scripts_scena_dat_us_t0050_dat
